COMPILER.md

```diff
@@ -0,0 +1,286 @@
+since these value distinctions are first-class in SExp, all values processed
+by the compiler, starting with the preprocessor (src/compiler/preprocessor.rs), 
+going through the frontend (src/compile/frontend.rs) and to code generation
+(src/compiler/codegen.rs) use these values.  Every expression read from the
+input stream has a Srcloc (src/compiler/srcloc.rs) which indicates where it
+was read from in the input, and as long as the srcloc is copied from form to
+form, it allows the compiler to maintain known associations between input and
+results.  For example, you can implement syntax coloring by simply using
+parse\_sexp (src/compiler/sexp.rs) to parse a file, run it through the
+preprocessor and frontend, and then decorate the locations present in the
+HelperForms accessible from the CompilerForm that results.  If an error is
+returned, it contains a srcloc that's relevant.
+
+You can break down the basics of how chialisp compilation functions like this:
+
+        let mut allocator = Allocator::new();
+        let mut symbol_table = HashMap::new();
+        let runner = Rc::new(DefaultProgramRunner::new());
+        let opts = Rc::new(DefaultCompilerOpts::new(&use_filename))
+            .set_optimize(do_optimize)
+            .set_search_paths(&search_paths)
+            .set_frontend_opt(dialect > 21);
+        let parsed = parse_sexp(Srcloc::start(&use_filename), file_content.bytes())?;
+        let compileform = frontend(opts.clone(), &parsed)?;
+        let generated_code = codegen(&mut allocator, runner, opts, &compileform, &mut symbol_table)?;
+        // generated_code.1 is an Rc<SExp> which contains the output
+        // code.
+
+Everything before code generation is uninteresting, but I'll note at a high level
+how functions on PrimaryCodegen function.
+
+codegen starts by running start_codegen to introduce each helper to the
+PrimaryCodegen and based on their types, bin them into the appropriate parts
+of the code generator to lookup later:
+
+    Defconstants are turned into mods and run, unlike classic chialisp.
+    The result is stored in the PrimaryCodegen's constant set.
+    
+    The new defconst form is evaluated by putting all Helpers into an Evaluator
+    and asking it to shrink the constant's expression (resulting in constant
+    folding).  The folded value must reduce to a constant, and if it does,
+    it's stored in the constant set.
+    
+    Defmacros are converted to programs and compiled using the
+    CompilerOpts' compile_program method.  These methods on CompilerOpts
+    provide this kind of recursion with an eye to allowing the generation
+    of program code to be used during compilation.  The resulting code
+    is stored in the macro set.
+    
+Next, let desugaring takes place (my intention is to lift this out of codegen to
+a separate pass).
+    
+Let desugaring inspects each defun (because they appear in the compiled code) for
```

Comment:
```suggestion
"Let" desugaring inspects each defun (because they appear in the compiled code) for
```

Reply:
An example here might be tremendously useful to help the reader shift context to what you're talking about (namely "let forms").

Reply:
thanks